Add Specific Service Page

whoojemaflip opened the pull request:
Implements the Specific Service Page. See commits for more specific context.

[Trello card](https://trello.com/c/AfsQXUa3/539-build-specific-service-page-5)

![screen shot 2016-11-18 at 11 00 14](https://cloud.githubusercontent.com/assets/3466862/20427945/5ca8ad2a-ad7e-11e6-98b1-5c47a0e44704.png)

Mobbed by @whoojemaflip @deborahchua @brenetic

whoojemaflip committed:
Copy usable code from v2 prototype

Commit:
Implement working specific service page

whoojemaflip committed:
Improve the rendering of the Specific Service page

This removes a couple of n+1 queries and tidies up some whitespace.

whoojemaflip committed:
Add feature spec for the /services/:service page

davidbasalla commented:
Sorry in advance for the pedantry, but prefer to have commit messages in the format of Active verb in present tense first - eg https://github.com/alphagov/styleguides/blob/master/git.md#structure

davidbasalla commented:
The [first commit](https://github.com/alphagov/local-links-manager/pull/76/commits/04eceb54ee3db0267e3dd8011d0b05283913f9f3) is a bit confusing in that adds specs changes which get removed in a later commit. It also refers to the old style tiers, so I suspect that commit was made before the Tier model refactoring was done. Also it does not include the switch over of the index page to show page. So maybe some rebasing/squashing is required?

brenetic commented:
@davidbasalla Done and done 🚀

davidbasalla commented on app/presenters/any_link_presenter.rb:
```diff
@@ -0,0 +1,92 @@
+class AnyLinkPresenter
```
Would it be worth adding unit tests for this class? Also we still have the `LinkPresenter` and `LocalAuthorityPresenter` which appear to do similar things. Is there a possibility of getting rid of any of these classes?

whoojemaflip replied:
I'm not convinced that unit-tests will add much value here. The code is getting exercised in the feature specs, which is where we test the rest of the view layer.  Open to counter-arguments of course.

davidbasalla commented on app/presenters/any_link_presenter.rb:
```diff
@@ -0,0 +1,92 @@
+
+    def initialize(link, with_service: nil, with_service_links_count: 0, context:, local_authority: nil)
+      super(link)
+      @with_service = with_service
```
Totally minor, but could the `with_` be removed or does it add any sort of special meaning here? Both for `with_service` and `with_service_links_count`

davidbasalla commented on app/presenters/any_link_presenter.rb:
```diff
@@ -0,0 +1,92 @@
+      super(link)
+      @with_service = with_service
+      @with_service_links_count = with_service_links_count
+      @context = context
```
Is there a better name for `context`? It's not immediately clear to me what that refers to.

h-lame replied:
`context` is an object that we can use in the presenter to call action view helper methods (in this case we use it to call the named route helpers).  It's usually a rails controller, or the action view template where the presenter is constructed.   Perhaps `view_context` or `template` would be a better name?

whoojemaflip replied:
renamed to `view_context` for clarity.

davidbasalla commented on app/presenters/any_link_presenter.rb:
```diff
@@ -0,0 +1,92 @@
+      case link_container
+      when Link
+        ForALink
+      when LocalAuthority
```
Does this code branch ever get called? It looks like we only use the `AnyLinkPresenter` class on the services show page for Links, but never for LocalAuthorities? If so, then we should remove this, and maybe revisit what this class does as opposed to the already existing LinkPresenter?

h-lame replied:
This whole presenter may have been brought over from the prototype prematurely.  In the prototype we wanted to present all links, including the home page link in the same list view, so it was useful to have a single object that would act appropriately given a `Link` or a `LocalAuthority`.  If we're not doing that anymore it might not be useful, and you could put the extra methods from here into `LinkPresenter`.

davidbasalla commented on app/views/services/show.html.erb:
```diff
@@ -0,0 +1,45 @@
+            <h3><%= link_to presented_authority.name, local_authority_path(presented_authority.slug) %></h3>
+          </th>
+        </tr>
+        <% first_link, *remaining_links = *links_by_authority %>
```
Could you explain what this line does? Not used to seeing `*remaining_links = *links_by_authority` style notation

h-lame replied:
This could just be `first_link, *remaining_links = links_by_authority` as the splat on `links_by_authority` is redundant.  The splat on `remaining_links` means to collect all the rest of the values.  E.g.:

``` ruby
first, rest = [1,2,3,4,5]
first #=> 1
rest #=> 2

first, *rest = [1,2,3,4,5]
first #=> 1
rest #=> [2,3,4,5]
```

brenetic replied:
Takes the first item from `links_by_authority` and returns it in `first_link`. Any remaining links are returned in `remaining_links`. @h-lame magic

whoojemaflip replied:
one of the finest bits of functional ruby I've seen in a long while. A++++ would copy Murray's prototype code again.

Commit:
PR Feedback

Attempts to simplify, by deleting as much code as possible.
Addresses concerns raised in the PR

whoojemaflip committed:
Add a CSS .wrap-words class to handle long URLs

brenetic commented:
@davidbasalla @issyl0 @emmabeynon This PR has been updated since the last comments. Ready for another viewing 🚀

davidbasalla commented:
Looks good! Wondering whether we can retire the `LinkPresenter` class, but it's still used for Interactions at the moment, so we can probably do this at a later point.
